File: array/290WordPattern.java
public class 290WordPattern {

    /**
     * Given a pattern and a string s, find if s follows the same pattern.
     * 
     * Input: pattern = "abba", s = "dog cat cat dog"
     * Output: true
     */
    public boolean wordPattern(String pattern, String s) {
        HashMap<String, Character> m1 = new HashMap<>();
        HashMap<Character, String> m2 = new HashMap<>();

        String[] words = s.split(" ");
        if(words.length != pattern.length()) return false;
        for(int i = 0; i < pattern.length(); i++) {
            if(!m1.containsKey(words[i])) {
                if(!m2.containsKey(pattern.charAt(i))) {
                    m1.put(words[i], pattern.charAt(i));
                    m2.put(pattern.charAt(i), words[i]);
                } else return false;
            } else {
                if(!m2.containsKey(pattern.charAt(i))) return false;
                if(m1.get(words[i]) != pattern.charAt(i)) return false;
            }
        }
        return true;
    }
}

File: company/capital_one/easy/290WordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        s1, s2 = dict(), dict()
        s = s.split(' ')
        if len(pattern) != len(s):
            return False
        
        for i in range(len(pattern)):
            if pattern[i] not in s1 and s[i] not in s2:
                s1[pattern[i]] = s[i]
                s2[s[i]] = pattern[i]
            elif pattern[i] not in s1 or s[i] not in s2:
                return False
            elif s1[pattern[i]] != s[i] or s2[s[i]] != pattern[i]:
                return False
            
        return True